File: leetcode-subscription/c#/Problems/P0170.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetCode.Naive.Problems
{
  /// <summary>
  ///    Problem: https://leetcode.com/problems/two-sum-iii-data-structure-design/
  ///    Submission: https://leetcode.com/submissions/detail/452797921/
  ///    LinkedIn
  /// </summary>
  internal class P0170
  {
    public class TwoSum
    {
      SortedList<(int value, long ticks), (int value, long ticks)> _s =
        new SortedList<(int value, long ticks), (int value, long ticks)>(new Comp());

      /** Initialize your data structure here. */
      public TwoSum()
      {

      }

      /** Add the number to an internal data structure.. */
      public void Add(int number)
      {
        var el = (number, DateTime.UtcNow.Ticks);

        _s.Add(el, el);
      }

      /** Find if there exists any pair of numbers which sum is equal to the value. */
      public bool Find(int value)
      {
        for (var i = 0; i < _s.Count; i++)
        {
          var j = BinarySearch(_s, value - _s.Keys[i].value);
          if (j == -1)
            continue;

          if (i == j)
            continue;

          return true;
        }

        return false;
      }

      private int BinarySearch(SortedList<(int value, long ticks), (int value, long ticks)> s, int limit)
      {
        var l = 0;
        var r = s.Count - 1;

        if (l == r)
        {
          if (_s.Keys[l].value == limit)
            return l;
          return -1;
        }

        while (true)
        {
          if (l + 1 == r)
          {
            if (_s.Keys[r].value == limit)
              return r;
            if (_s.Keys[l].value == limit)
              return l;
            return -1;
          }

          var mid = (l + r) / 2;
          if (_s.Keys[mid].value >= limit)
            r = mid;
          else
            l = mid;
        }
      }

      public class Comp : IComparer<(int value, long ticks)>
      {
        public int Compare((int value, long ticks) x, (int value, long ticks) y) =>
          x.value == y.value ? x.ticks.CompareTo(y.ticks) : x.value.CompareTo(y.value);
      }
    }
  }
}

File: binarysearch/src/main/java/binarysearch/P0170.java
package binarysearch;

/*
 * Problem: https://binarysearch.com/problems/Number-of-Unique-Binary-Search-Trees
 * Submission: https://binarysearch.com/problems/Number-of-Unique-Binary-Search-Trees/submissions/7380503
 */
public class P0170 {
  class Solution {
    public int solve(int n) {
      var memo = new int[n + 1];
      return count(n, memo);
    }

    private int count(int n, int[] memo) {
      if (n == 0)
        return 1;

      if (memo[n - 1] != 0)
        return memo[n - 1];

      var sum = 0;
      var mod = (int)1e9 + 7;

      for (var i = 0; i < n; i++) {
        if (i == 0 || i == n - 1) {
          sum += count(n - 1, memo);
        }
        else {
          long value = ((long)count(i, memo)) * count(n - i - 1, memo);
          sum += (value % mod);
        }
        sum %= mod;
      }

      memo[n - 1] = sum;
      return sum;
    }
  }
}